drivers/can/can_renesas_rz_canfd.c

```diff
@@ -0,0 +1,1033 @@
+	 BIT(R_CANFD_CFDC_CTR_EPIE_Pos) |  /* Error Passive Interrupt Enable */                    \
+	 BIT(R_CANFD_CFDC_CTR_BOEIE_Pos) | /* Bus-Off Entry Interrupt Enable */                    \
+	 BIT(R_CANFD_CFDC_CTR_BORIE_Pos) | /* Bus-Off Recovery Interrupt Enable */                 \
+	 BIT(R_CANFD_CFDC_CTR_OLIE_Pos))   /* Overload Interrupt Enable */
+
+#define RULE_NUM_PER_PAGE (16)
+
+#define CANFD_CFG_RXFIFO_DEFAULT(x)                                                                \
+	((CANFD_CFG_RXFIFO##x##_INT_THRESHOLD << R_CANFD_CFDRFCC_RFIGCV_Pos) |                     \
+	 (CANFD_CFG_RXFIFO##x##_DEPTH << R_CANFD_CFDRFCC_RFDC_Pos) |                               \
+	 (CANFD_CFG_RXFIFO##x##_PAYLOAD << R_CANFD_CFDRFCC_RFPLS_Pos) |                            \
+	 (CANFD_CFG_RXFIFO##x##_INT_MODE) | (CANFD_CFG_RXFIFO##x##_ENABLE))
+
+/* Default Dataphase bitrate configuration in case classic mode is enabled */
+static const can_bit_timing_cfg_t classic_can_data_timing_default = {
+	.baud_rate_prescaler = 1,
+	.time_segment_1 = 3,
+	.time_segment_2 = 2,
+	.synchronization_jump_width = 1,
+};
+
+struct can_renesas_rz_filter {
+	bool set;
+	struct can_filter filter;
+	can_rx_callback_t rx_cb;
+	void *rx_usr_data;
+};
+
+struct can_renesas_rz_cfg {
+	struct can_driver_config common;
+	const struct pinctrl_dev_config *pin_config;
+	const struct device *global_dev;
+	const uint32_t rx_filter_num;
+	const can_api_t *fsp_api;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	uint32_t clock_ext_rate;
+};
+
+struct can_renesas_rz_data {
+	struct can_driver_data common;
+	struct k_mutex inst_mutex;
+	can_tx_callback_t tx_cb;
+	struct k_sem tx_sem;
+	void *tx_usr_data;
+	struct can_renesas_rz_filter *rx_filter;
+	can_bit_timing_cfg_t data_timing;
+
+	canfd_instance_ctrl_t *fsp_ctrl;
+	can_cfg_t *fsp_cfg;
+};
+
+extern void canfd_error_isr(void);
+extern void canfd_rx_fifo_isr(void);
+extern void canfd_channel_tx_isr(void);
+static inline int can_renesas_rz_apply_default_config(const struct device *dev);
+
+static inline int set_hw_timing_configuration(const struct device *dev,
+					      can_bit_timing_cfg_t *f_timing,
+					      const struct can_timing *z_timing)
+{
+	struct can_renesas_rz_data *data = dev->data;
+
+	if (f_timing == NULL || z_timing == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	*f_timing = (can_bit_timing_cfg_t){
+		.baud_rate_prescaler = z_timing->prescaler,
+		.time_segment_1 = z_timing->prop_seg + z_timing->phase_seg1,
+		.time_segment_2 = z_timing->phase_seg2,
+		.synchronization_jump_width = z_timing->sjw,
+	};
+
+	k_mutex_unlock(&data->inst_mutex);
+
+	return 0;
+}
+
+static inline int get_free_filter_id(const struct device *dev)
+{
+	struct can_renesas_rz_data *data = dev->data;
+	const struct can_renesas_rz_cfg *cfg = dev->config;
+	int ret = 0;
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
```

Comment:
Won't cause issues but the Single caller using this already has locked the instance, so no need to lock it a second time.